packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -1844,6 +1674,219 @@ class FocusManager with DiagnosticableTreeMixin, ChangeNotifier {
   }
 }
 
+// A class to detect and manage the highlight mode transitions.
+// An instance of this is owned by the FocusManager.
+class _HighlightModeManager {
+  _HighlightModeManager();
+
+  // If set, indicates if the last interaction detected was touch or not.
+  // If null, no interactions have occurred yet.
+  bool? _lastInteractionWasTouch;
+
+  FocusHighlightMode get highlightMode => _highlightMode ?? _defaultModeForPlatform;
+  FocusHighlightMode? _highlightMode;
+
+  FocusHighlightStrategy get strategy => _strategy;
+  FocusHighlightStrategy _strategy = FocusHighlightStrategy.automatic;
+  set strategy(FocusHighlightStrategy value) {
+    if (_strategy == value) {
+      return;
+    }
+    _strategy = value;
+    updateMode();
+  }
+
+  /// Register a closure to be called when the [FocusManager] notifies its listeners
+  /// that the value of [highlightMode] has changed.
+  void addListener(ValueChanged<FocusHighlightMode> listener) => _listeners.add(listener);
+
+  /// Remove a previously registered closure from the list of closures that the
+  /// [FocusManager] notifies.
+  void removeListener(ValueChanged<FocusHighlightMode> listener) => _listeners.remove(listener);
```

Comment:
Ooh, good catch. If I could have made the empty one const, I would have. Instead of the static empty list, I decided to just re-init the _listeners to a new empty list when it gets disposed, so that there's no chance that something will operate on a static list, inserting things while disposed, making all disposed instances point to the same (now non-empty!) list.